pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,106 @@
+package writelimiter
+
+import (
+	"runtime"
+	"sync/atomic"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+type blobWriter struct {
+	distribution.BlobWriter
+	release func()
+}
+
+var _ distribution.BlobWriter = blobWriter{}
+
+func (bw blobWriter) Close() error {
+	bw.release()
+	return bw.BlobWriter.Close()
+}
+
+type blobStore struct {
+	distribution.BlobStore
+	sem chan struct{}
+}
+
+var _ distribution.BlobStore = blobStore{}
+
+func (bs blobStore) acquire() func() {
+	// LOCK_BEGIN++
+	sem := bs.sem
+	// LOCK_WAIT_BEGIN++
+	sem <- struct{}{}
+	// LOCK_WAIT_END++
+	done := int32(0)
+	return func() {
+		if atomic.SwapInt32(&done, 1) == 0 {
+			<-sem
+			// LOCK_END++
+		}
+	}
+}
+
+func (bs blobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	release := bs.acquire()
+
+	bw, err := bs.BlobStore.Create(ctx, options...)
+	if err != nil {
+		release()
+		return bw, err
+	}
+
+	blobwriter := blobWriter{
+		BlobWriter: bw,
+		release:    release,
+	}
+
+	// We must be sure that the lock is released even if Close() has not been called.
+	// Otherwise, a leak could happen.
+	runtime.SetFinalizer(blobwriter, func(*blobWriter) { release() })
```

Comment:
looks like some Cancel() implementations return without calling Close() if they encounter an error... escalating a resource leak to a lock leak is concerning

https://github.com/openshift/origin/blob/master/vendor/github.com/docker/distribution/registry/storage/blobwriter.go#L94